src/udp.rs

```diff
@@ -39,103 +114,293 @@ impl UdpSocket {
-        // TODO ensure write is currently possible
-        self.socket
-            .set_tx_read_pointer(bus, 0)
-            .and_then(|_| bus.write_frame(self.socket.tx_buffer(), 0, send_buffer))
-            .and_then(|_| {
-                self.socket
-                    .set_tx_write_pointer(bus, send_buffer.len() as u16)
-            })
-            .and_then(|_| self.socket.command(bus, socketn::Command::Send))?;
+        // Send the data.
+        self.socket.command(bus, socketn::Command::Send)?;
 
         loop {
-            if self.socket.get_tx_read_pointer(bus)? == self.socket.get_tx_write_pointer(bus)? {
-                if self.socket.has_interrupt(bus, socketn::Interrupt::SendOk)? {
-                    self.socket
-                        .reset_interrupt(bus, socketn::Interrupt::SendOk)?;
-                    return Ok(());
-                } else if self
-                    .socket
-                    .has_interrupt(bus, socketn::Interrupt::Timeout)?
-                {
-                    self.socket
-                        .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
-                    return Err(NbError::Other(UdpSocketError::WriteTimeout));
-                }
+            match self.try_flush_tx(bus) {
+                Err(NbError::WouldBlock) => {}
+                result => return result,
             }
         }
     }
 
+    fn try_flush_tx<SpiBus: Bus>(
+        &self,
+        bus: &mut SpiBus,
+    ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
+        let tx_read = self.socket.get_tx_read_pointer(bus)?;
+        let tx_write = self.socket.get_tx_write_pointer(bus)?;
+
+        if tx_read == tx_write {
+            if self.socket.has_interrupt(bus, socketn::Interrupt::SendOk)? {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::SendOk)?;
+
+                return Ok(());
+            } else if self
+                .socket
+                .has_interrupt(bus, socketn::Interrupt::Timeout)?
+            {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
+
+                return Err(NbError::Other(UdpSocketError::WriteTimeout));
+            }
+            // other interrupts are unreachable in UDP mode.
+
+            return Ok(());
+        }
+
+        Err(NbError::WouldBlock)
+    }
+
     /// Sets a new destination before performing the send operation.
+    ///
+    /// # Returns
+    ///
+    /// The amount of bytes that were sent. The caller should make sure to
+    /// check and send the rest of the `send_buffer` data.
     fn socket_send_to<SpiBus: Bus>(
         &mut self,
         bus: &mut SpiBus,
         remote: SocketAddrV4,
         send_buffer: &[u8],
     ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
         self.set_destination(bus, remote)?;
-        self.socket_send(bus, send_buffer)
+
+        self.socket_send_all(bus, send_buffer)
```

Comment:
Do we need to flush the TX buffer before setting the destination?

Reply:
If someone hits a problem with this in the future, they can submit a PR. In general, I'm concerned that someone calls a non-blocking `send()` followed by a `socket_send_to()`. This means there'd be some data ready for one dest in the device when we swap over the destination.